src/message_broker.cpp

```diff
@@ -0,0 +1,136 @@
+#include "message_broker.h"
+
+#include "websocket/websocket_endpoint.h"
+#include "websocket/websocket_secure_endpoint.h"
+
+namespace sensr {
+
+MessageBrokerBase::MessageBrokerBase(const std::string& address,
+                                     uint16_t port,
+                                     const std::string& cert_path) {
+  if (cert_path.empty()) {
+    endpoint_ = std::make_unique<WebSocketEndPoint>(address, port);
+  } else {
+    endpoint_ = std::make_unique<WebSocketSecureEndPoint>(address, port, cert_path);
+  }
+}
+
+bool MessageBrokerBase::IsConnected() const { return endpoint_->IsConnected(); }
+
+bool MessageBrokerBase::IsListening() const { return !listeners_.empty(); }
+
+void MessageBrokerBase::Reconnect() {
+  if (listeners_.empty()) {
+    return;
+  }
+
+  endpoint_->Close(websocketpp::close::status::normal);
+
+  using namespace std::chrono_literals;
+  std::this_thread::sleep_for(1ms);
+
+  endpoint_->Connect();
+}
+
+void MessageBrokerBase::TryAttachListener(const std::shared_ptr<MessageListener>& listener) {
+  bool start_connection = listeners_.empty();
+
+  if (listener != nullptr && IsAttachable(*listener)) {
+    if (auto it = std::find(listeners_.cbegin(), listeners_.cend(), listener);
+        it == listeners_.cend()) {
+      listeners_.push_back(listener);
+    }
+  }
+
+  if (start_connection && !listeners_.empty()) {
+    endpoint_->Connect();
+  }
+}
+
+void MessageBrokerBase::TryDetachListener(const std::shared_ptr<MessageListener>& listener) {
+  bool end_connection = !listeners_.empty();
+
+  if (listener != nullptr && IsAttachable(*listener)) {
+    if (auto it = std::find(listeners_.cbegin(), listeners_.cend(), listener);
+        it != listeners_.cend()) {
+      listeners_.erase(it);
+    }
+  }
+
+  if (end_connection && listeners_.empty()) {
+    endpoint_->Close(websocketpp::close::status::normal);
+  }
+}
+
+void MessageBrokerBase::ClearListeners() {
+  listeners_.clear();
+  endpoint_->Close(websocketpp::close::status::normal);
+}
+
+OutputMessageBroker::OutputMessageBroker(const std::string& address,
+                                         uint16_t port,
+                                         const std::string& cert_path)
+    : MessageBrokerBase(address, port, cert_path) {
+  endpoint_->SetCallbacks([this](auto&& msg) { OnMessage(msg); },
+                          [this](auto&& err) { OnError(err); });
```

Comment:
I intended to not virtualize `OnMessage` and `OnError` because virtual functions might not be initialized in the constructor. https://stackoverflow.com/a/962148